Don't delete misplaced items on final attempt

arbrandes opened the pull request:
## Description

In assessment mode, don't delete misplaced items from the user's state when handling the final attempt.

This is a server-side change to match what Javascript already did.  In other words, upon submitting the final attempt, misplaced items were already left on the board.  However, since the corresponding state was deleted server-side, upon refreshing the page the items were returned to the bank.

## Dependencies

*None*

## JIRA

*None*

## Sandbox URLs

- [LMS](http://dndv2-sol-1998.opencraft.hosting/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2/)
- [Studio](http://studio-dndv2-sol-1998.opencraft.hosting/container/block-v1:test+tt101+run+type@vertical+block@drag_and_drop_v2)

## Latest commit

- 479b8dd *(updated 2016-09-27)*

## Testing instructions

1. Go to the [pre-deployed DnDv2 instance](http://dndv2-sol-1998.opencraft.hosting/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2/) on the LMS, and log in as `staff@example.com`.
2. Click on `Staff Debug Info`, and `Delete Student State` for the `staff@example.com` user.
3. Close the debug info window, and drag a single item into its correct zone (for instance, "Goes to the Top" to the top zone).
4. Click `Submit` twice, so that you only have one attempt left.
5. Drag an item into an incorrect zone (for instance, drag "I don't belong anywhere" to the bottom zone).
6. Click `Submit`.
7. Close the incorrect answer popup.
8. Refresh the page.
9. Notice that even though you have no attempts left, the wrongly placed item is still there.

## Screenshots

### Wronly placed item is still there after refreshing the page.

![screenshot](https://cloud.githubusercontent.com/assets/759355/18894195/00ed7ad8-84e8-11e6-952a-718a9dec6606.png)

arbrandes commented:
@itsjeyd, @bradenmacdonald, should I include this in #10?

bradenmacdonald commented:
@arbrandes We would need to pull it into https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/101 and I'll let @itsjeyd decide. If upstream hasn't gotten into code review yet then it would make sense to include imho.

itsjeyd commented:
@arbrandes @bradenmacdonald We seem to be getting close to merging https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/101 -- it's missing a review by Christina, but Albert is pretty much done reviewing and Shelby has also given the PR a pass. So I think I'd rather not complicate things by extending the scope of https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/101. 

However, it would be good to get this fix in quickly. @bradenmacdonald If we can create a separate ticket for reviewing this PR, I'll take it on for next sprint and review this PR tomorrow.

arbrandes commented:
Note: I can't reproduce the build failure locally.

itsjeyd commented on tests/integration/test_interaction_assessment.py:
```diff
@@ -130,6 +130,35 @@ def test_misplaced_items_returned_to_bank(self):
+    def test_misplaced_items_not_returned_to_bank_on_final_attempt(self):
+        """
+        Test items placed to incorrect zones are not returned to item bank
+        after submitting solution on the final attempt
```
@arbrandes This test also checks what happens when refreshing the page; it would be worth mentioning that here I think.

itsjeyd commented on tests/unit/test_advanced.py:
```diff
@@ -390,6 +390,17 @@ def test_do_attempt_shows_final_feedback_at_last_attempt(self):
         expected_message = self._make_feedback_message(self.FINAL_FEEDBACK)
         self.assertIn(expected_message, res[self.OVERALL_FEEDBACK_KEY])
 
+    def test_do_attempt_does_not_delete_misplaced_items_at_last_attempt(self):
```
@arbrandes Could you add a docstring to this test please?

itsjeyd commented:
@arbrandes This looks like a really nice patch! I only had a couple of comments regarding docstrings.

For manual testing we can reuse an existing DnDv2 [sandbox](http://dndv2-pr96.opencraft.hosting/). You should have access to it since it was created for one of your PRs that already got merged. (Let me know if that's not the case for some reason.) Could you install this branch on the sandbox and set up a test course so I can verify the changes there? Thanks!

itsjeyd commented:
@arbrandes I updated `open-craft:master`

itsjeyd commented:
Hi @arbrandes, were you able to get these changes set up on the sandbox for testing? Just checking in to see where you're at with this :)

arbrandes commented:
@itsjeyd, forgive the hiccup: I've updated the PR description (it wasn't up to snuff), rebased on master, pushed a fixup addressing your comments, and updated the SOL-1998 sandbox with 479b8dd.

As for thanking me for my patience on the Show Answer PR, it was my pleasure!  After all, the tougher the job, the greater the reward!  Seeing it merged made me all warm and fuzzy inside.  :)

itsjeyd commented:
@arbrandes That's great to hear :)

Nice work on this! I tested the changes on the sandbox, and everything is working as expected. You didn't mention it in the PR description, but I found that the changes also integrate well with the "Show Answer" feature. (Last attempt is restored when refreshing the page after clicking "Show Answer".)

:+1:

- [x] I tested this: Works as advertised in PR description
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ This is a bug fix that doesn't introduce new workflows (or alter existing ones).
- [ ] ~~Includes documentation~~ This is a bug fix that doesn't require documentation updates.

Upstream PR for this is here: https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/106.

itsjeyd commented:
@cahrens @staubina This is ready for you to have a look. Sandbox links are in the PR description. Thanks!

itsjeyd commented:
@cahrens @staubina Added the previous [comment](https://github.com/open-craft/xblock-drag-and-drop-v2/pull/13#issuecomment-250110188) to the wrong PR. Please head over to https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/106. Sorry for the confusion.
